walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,114 @@
+    {
+      return Arrays.asList();
+    }
+
+    // Return full day for trivial empty cases (no event given, no attendees given)
+    if (events.isEmpty() || request.getAttendees().isEmpty())
+    {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    events = sortEventsByStartTime(events);
+
+    int requiredDuration = (int)request.getDuration(); // Can cast as such because duration does not exceed 2^32.
+    Collection<String> requiredAttendees = request.getAttendees();
+
+    ArrayList<TimeRange> candidateRanges = new ArrayList<TimeRange>();
+
+    Iterator eventsIterator = events.iterator();
+    Event currentEvent = (Event)eventsIterator.next();
+    Event previousEvent = currentEvent;
+    int currentTime = TimeRange.START_OF_DAY;
+    int startTime = TimeRange.START_OF_DAY;
+    
+    while (true) {
+
+      boolean attendeesNotIncludedInCurrentEvent = Collections.disjoint(currentEvent.getAttendees(), request.getAttendees());
+      
+      if (attendeesNotIncludedInCurrentEvent == false) 
+      { 
+        currentTime = currentEvent.getWhen().start(); // Available time is only up to when event starts, since there are clashing attendees.
+
+        TimeRange potentialSlot = TimeRange.fromStartEnd(startTime, currentTime, false);
+        if (potentialSlot.duration() >= requiredDuration)
+        {
+          candidateRanges.add(potentialSlot);
+        }
+
+       startTime = setStartTimeBasedOnPreviousEvent(previousEvent, currentEvent, request);
+      }
+
+      // Check for next event
+      if (eventsIterator.hasNext())
+      {
+        previousEvent = currentEvent;
+        currentEvent = (Event)eventsIterator.next();
+        continue;
+      }
+      else // Terminate
+      {
+        TimeRange potentialSlot = TimeRange.fromStartEnd(startTime, TimeRange.END_OF_DAY, true);
+        if (potentialSlot.duration() >= requiredDuration)
+        {
+          candidateRanges.add(potentialSlot);
+        }
+      
+        break;
+      }
+      
+    }
+
+    return candidateRanges;
+  }
+
+  public Collection<Event> sortEventsByStartTime(Collection<Event> events)
+  {
+    List<Event> list = new ArrayList<Event>(events);
+
+    Collections.sort(list, (a, b) -> TimeRange.ORDER_BY_START.compare(a.getWhen(), b.getWhen()));
+    return list;
+  }
+
+  private int setStartTimeBasedOnPreviousEvent(Event prevEvent, Event currEvent, MeetingRequest request)
+  {
+    boolean previousEventIsOverlapping = prevEvent != currEvent && prevEvent.getWhen().overlaps(currEvent.getWhen());
+    boolean attendeesNotIncludedInPreviousEvent = Collections.disjoint(prevEvent.getAttendees(), request.getAttendees());
+
+    if (attendeesNotIncludedInPreviousEvent)
+    {
+      return currEvent.getWhen().end();
+    }
+
+    if (previousEventIsOverlapping) {
+      return Math.max(currEvent.getWhen().end(),prevEvent.getWhen().end());
+    }
+
+    // Not overlapping, previous event is irrelevant
+    return currEvent.getWhen().end();    
```

Comment:
You can simplify this to: 
`return Math.max(currEvent.getWhen().end(),prevEvent.getWhen().end());`

If the previous event is non-overlapping, then naturally the end time will be before the current event end time.

Reply:
*smacks forehead* of course. thanks for spotting that!